OS and Process resource detectors

nelsonghezzi opened the pull request:
Hi,

This PR adds detectors for most of the attributes under the `OS` and `Process` resources.

The only not implemented are:
* `process.command`
* `process.command_line`
* `os.description`

New functions are provided under the `resource` package:
* `resource.WithOSType()`
* `resource.WithProcess()`
* `resource.WithProcessPID()`
* `resource.WithProcessExecutableName()`
* `resource.WithProcessExecutablePath()`
* `resource.WithProcessCommandArgs()`
* `resource.WithProcessOwner()`
* `resource.WithProcessRuntimeName()`
* `resource.WithProcessRuntimeVersion()`
* `resource.WithProcessRuntimeDescription()`

Tests are `TODO` atm. Will add them if you think the PR it's a good fit for the package.

Some output examples:

Linux (container):
<img width="1076" alt="image" src="https://user-images.githubusercontent.com/6424021/114123203-6127e480-98e1-11eb-9592-86330f01df2f.png">

Darwin:
<img width="1076" alt="image" src="https://user-images.githubusercontent.com/6424021/114123251-74d34b00-98e1-11eb-9ee7-3b434e62008c.png">

Windows:
<img width="709" alt="image" src="https://user-images.githubusercontent.com/6424021/114123263-7b61c280-98e1-11eb-9c7d-b0d5c544a7df.png">

Aneurysm9 commented:
I think this seems like a reasonable way to add detection of these resources.  Please add some tests and a `CHANGELOG.md` entry.

nelsonghezzi commented:
@Aneurysm9 Done! Let me know if it's more or less what you had in mind.

Aneurysm9 commented:
I worry somewhat that the tests are just restatements of the logic being tested.  Given the nature of these resource detectors though, I'm not sure there's a good way to mock up the environment for testing other than perhaps creating an interface and a struct with proxy methods hanging off of it so that you can mock that interface.  Is that worth the effort here? Do we expect `runtime` to be a problem? v0v

nelsonghezzi commented:
I had the same impression while writing the tests. I also thought about the interface/mock approach, as I have done it before to test functions with dependencies on `rand`, but ended up thinking it was not worth it in this case.

There is a reason why I replicated the logic in the test file (the `pid()`, `executableName()` and similar functions).

I could have defined those functions in the `process.go` file and then exported them for reuse in the tests (the `export_test.go` trick). But doing so would only have reinforced the feeling of self-validating code.

Instead my reasoning was "I assume the `os`, `runtime` and `user` packages return the expected (and correct) information for these attributes, so this is the information that the `WithProcess*` functions should set on each of them".

Besides that (which was more of a conceptual relief I've to admit 😅), the tests only validates that each function sets the expected attribute name (with the exception of the runtime description test maybe, that also validates the string format).

My concern is mostly in the command args test, because is the most susceptible to what the test runner put on them. `os.Args` is writeable, but that would be a risky way to mock them, as it could affect subsequent tests if not restored I think.

I will search for other ways to mock the environment, but so far didn't find anything.

nelsonghezzi commented:
I found [this question](https://forum.golangbridge.org/t/how-to-mock-os-package-function-calls-in-a-test-case-at-go/8562), which gave me an idea for a possible lightweight form of indirection, allowing to mock the expected attribute values.

I've added 886314ccd20e75190967fde2e5a1cd2e48b9f94f to further explore the idea.

But if you think the extra complexity doesn't pay for the benefits, I can revert to the previous version.

Aneurysm9 commented:
The approach of having local function variables that can be replaced in tests could work.  I think I'd prefer to see those function variables only reference the `os` or `runtime` functions or variables and not contain formatting logic.  The formatting logic should still live in the detector functions and thus be testable by altering inputs and asserting different outputs.

For instance, rather than have:

```go
	executableName executableNameProvider = func() string { return filepath.Base(os.Args[0]) }
	commandArgs    commandArgsProvider    = func() []string { return os.Args }
...
// Detect returns a *Resource that describes the name of the process executable.
func (processExecutableNameDetector) Detect(ctx context.Context) (*Resource, error) {
	return NewWithAttributes(semconv.ProcessExecutableNameKey.String(executableName())), nil
}
```

I would have:

```go
	commandArgs    commandArgsProvider    = func() []string { return os.Args }
...
// Detect returns a *Resource that describes the name of the process executable.
func (processExecutableNameDetector) Detect(ctx context.Context) (*Resource, error) {
	return NewWithAttributes(semconv.ProcessExecutableNameKey.String(filepath.Base(commandArgs()[0]))), nil
}
```

That way you can set up a table-based test for that detector that tests various inputs and ensures it produces the expected output:

```
/foo/bar/baz => baz
foo => foo
etc.
```

nelsonghezzi commented:
That makes sense.

I've streamlined the initial implementation based on your feedback, let me know what you think.

These are the main changes:
* Splitted functions to set "providers" based on the package the wrapped functions/variables comes from (I think this could allow for more granular reuse in the OS detector tests).
* Added a way to restore "default" implementations for those functions (mostly for use on the tear down phase of the tests, I think it's safe if the tests don't leave this variables changed).
* Added tests to exercise the error code paths, as this is easy now.
* Added tests to verify the provider functions that wrap variables (this was mostly to not impact the coverage leaving those paths untested, but they don't add much value otherwise).

p.s.: I think the CI is failing randomly. Is there a way to manually trigger a new run?

nelsonghezzi commented:
Hi! I'm keeping this PR up to date with `main`.

If you think the current approach ended up a bit twisted, maybe we can revert back to the more "naive" way of testing, and iterate the mocked approach in a separate PR?

I also have a cross platform implementation for the `os.description` attribute, but it will be better to submit it in a separate PR to keep this in a manageable size. Because it's partially based on the work made here, I'm holding it for now.

MrAlias commented on CHANGELOG.md:
```diff
@@ -37,6 +37,16 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - The Jaeger exporter now reports dropped attributes for a Span event in the exported log. (#1771)
 - Adds `k8s.node.name` and `k8s.node.uid` attribute keys to the `semconv` package. (#1789)
 - Adds `otlpgrpc.WithTimeout` option for configuring timeout to the otlp/gRPC exporter. (#1821)
+- Added `WithOSType` resource configuration option to set OS (Operating System) type resource attribute (`os.type`).
+- Added `WithProcess*` resource configuration options to set Process resource attributes.
```
```suggestion
- Added `WithOSType` resource configuration option to set OS (Operating System) type resource attribute (`os.type`). (#1788)
- Added `WithProcess*` resource configuration options to set Process resource attributes. (#1788)
```

MrAlias commented on sdk/resource/os.go:
```diff
@@ -0,0 +1,38 @@
+	), nil
+}
+
+// WithOS adds an attribute with the operating system type to the configured Resource.
```
```suggestion
// WithOSType adds an attribute with the operating system type to the configured Resource.
```

MrAlias commented on sdk/resource/os.go:
```diff
@@ -0,0 +1,38 @@
+// service is running on.
+func (osTypeDetector) Detect(ctx context.Context) (*Resource, error) {
+	return NewWithAttributes(
+		semconv.OSTypeKey.String(strings.ToUpper(runtime.GOOS)),
```
[Based on the current specification](https://github.com/open-telemetry/opentelemetry-specification/blob/b4ebec7e272fc6b6ad31a47485c91a2f7a2534d9/specification/resource/semantic_conventions/os.md) these need to be lowercase.

nelsonghezzi replied:
Oh, good catch! It seems like it just [changed](https://github.com/open-telemetry/opentelemetry-specification/commit/b4ebec7e272fc6b6ad31a47485c91a2f7a2534d9#diff-9ecbdd017d5e18e6b7ccbf1c82530db4fbdc07032d0fbdcc5a13e658bb0dd2c7) last week.

nelsonghezzi replied:
Addressed in ac97d34c17109d7813d073e9bf16a2d51e8ea641.

Although the returned string in `runtime.GOOS` is already lowercase, I left an explicit `ToLower` for it be future-proof just in case the current casing changes.

MrAlias commented on sdk/resource/process.go:
```diff
@@ -0,0 +1,237 @@
+	return WithDetectors(processCommandArgsDetector{})
+}
+
+// WithProcessOwner adds an attribute the username of the user that owns the process
```
```suggestion
// WithProcessOwner adds an attribute with the username of the user that owns the process
```

MrAlias reviewed (approved):
I think these detectors are a great idea, thanks for the contribution. I'm a bit worried about the tests being more change detectors than regression tests as was pointed out. I'm good with accepting them as is though and if we determine them to be truly deficient we can address then.

:1st_place_medal:

nelsonghezzi commented:
Hi @MrAlias, thanks for the feedback! I think I solved all the points, but let me know otherwise (don't know why the CI stalled with the latests commits).

Yeah, I also feel that the tests could be expanded and/or improved somehow. Maybe as the codebase evolves (and some other resource detectors are added) a more common pattern emerges. At that point, count on me to revisit this implementation :)
